### PR TITLE
Switch to swift calling conv

### PR DESCRIPTION
**Pull request by aschwaighofer**
Switch over to using the swift calling convention for swift functions

- Use the generic type lowering algorithm described in
"docs/CallingConvention.rst#physical-lowering" to map from IRGen's explosion
type to the type expected by the Swift ABI.

- Change IRGen to use the swift calling convention (swiftcc) for native swift
functions.
  Use the 'swiftself' attribute on self parameters and for closures contexts.
  Use the 'swifterror' parameter for swift error parameters.

- Change functions in the runtime that are called as native swift functions to use
the swift calling convention.

rdar://problem/19979055

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
I think this deserves a full test!

**Comment by aschwaighofer** [comment]
@jrose-apple Totally agree :-)

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test linux platform

**Comment by aschwaighofer** [comment]
@jrose-apple There is some local testing that I want to do before this should go in. At this time I will at the latest also run full test.

**Comment by gottesmm** [comment]
In fact, I would even say a full clean test since you are making some cmake changes.

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test linux platform

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test os x platform

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test linux

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test os x platform

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test linux platform

**Comment by aschwaighofer** [comment]
@swift-ci Please clean test

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test os x platform

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test linux platform

**Comment by jrose-apple** [comment]
Did this ever pass the full tests? I assume so because there were no comments after the Wrong SHA failures, but it'd be nice to know for sure.

**Comment by jrose-apple** [comment]
Also, were there any major benchmark changes from this?

**Comment by aschwaighofer** [comment]
Yes it passed the full test:

https://ci.swift.org/job/swift-PR-osx/5542/
https://ci.swift.org/job/swift-PR-Linux/5575/

Unfortunately, I had to rebase after that.

**Comment by aschwaighofer** [comment]
I did not measure significant changes in local testing.

**Comment by jrose-apple** [comment]
Looks like this broke the optimized tests for the simulator: https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RD_test-simulator/2738/

```
******************** TEST 'Swift(macosx-x86_64) :: Sanitizers/tsan-ignores-arc-locks.swift' FAILED ********************
Script:
--
xcrun --toolchain default --sdk /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RD_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swiftc -target x86_64-apple-macosx10.9  -module-cache-path '/var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/swift-testsuite-clang-module-cachepeMyEm' -F /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/../../../Developer/Library/Frameworks -Xlinker -rpath -Xlinker /Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/../../../Developer/Library/Frameworks  -swift-version 3 -O -sanitize=thread /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RD_test-simulator/swift/test/Sanitizers/tsan-ignores-arc-locks.swift -o /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RD_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/test-macosx-x86_64/Sanitizers/Output/tsan-ignores-arc-locks.swift.tmp_binary
TSAN_OPTIONS=ignore_interceptors_accesses=1:halt_on_error=1 /Users/buildnode/jenkins/workspace/oss-swift_tools-RA_stdlib-RD_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/test-macosx-x86_64/Sanitizers/Output/tsan-ignores-arc-locks.swift.tmp_binary
--
Exit Code: 1

Command Output (stderr):
--
swifterror value can only be loaded and stored from, or as a swifterror argument!
%swift.error** %3
  %103 = bitcast %swift.error** %3 to i8*
LLVM ERROR: Broken function found, compilation aborted!
```

**Comment by aschwaighofer** [comment]
Taking a look

**Comment by gottesmm** [comment]
@shahmishal @jrose-apple I think the os x smoke tests are fast enough that we should consider running the optimized tests on OS X.

Thoughts?

**Comment by jrose-apple** [comment]
Thoughts: "a PR is the wrong place to have this discussion". :-)

**Comment by jrose-apple** [comment]
Arnold also ran a regular set of tests. Those don't do optimized either.

**Comment by gottesmm** [comment]
@jrose-apple Sure. I was more just putting it out there before I send out an email

**Comment by aschwaighofer** [comment]
:( I think the thread sanitizer inserts uses of the swifterror value. I will have to fix that.

**Comment by aschwaighofer** [comment]
Upstream patch in the works ...

**Comment by shahmishal** [comment]
@gottesmm https://github.com/apple/swift/pull/2476 I was trying to enable it for Linux, however we have few test failing.

**Comment by aschwaighofer** [comment]
llvm/stable picked up the fix. Should be fixed as soon as the failing bot picks up the fix.

File: lib/IRGen/GenFunc.cpp [hunk 1]
```diff
@@ -1301,7 +1331,12 @@ void irgen::emitFunctionPartialApplication(IRGenFunction &IGF,
   // to capture), and the dest ownership semantics match the parameter's,
   // skip building the box and thunk and just take the pointer as
   // context.
-  if (!origType->isPolymorphic() &&
+  // TODO: We can only do this and use swiftself if all our swiftcc emit the
+  // last parameter that fits into a register as swiftself.
+  // We should get this optimization back using the @convention(closure) whose
+  // box argument should just be swift self.
+  if (false &&
```
Comment (rudkx) [inline lib/IRGen/GenFunc.cpp]: This is resulting in a build warning about unreachable code.

Can you (ideally) remove this code and open a bug to restore the behavior that you'd like to eventually see, or alternately follows he advice in the note we emit (to add parens around false to silence the warning).

Comment (aschwaighofer) [inline lib/IRGen/GenFunc.cpp]: https://github.com/apple/swift/pull/7702
https://bugs.swift.org/browse/SR-4041

**Pull request merged**